components/dashboard/src/settings/Plans.tsx

```diff
@@ -4,13 +4,462 @@
+                        {!Plans.isFreePlan(confirmDowngradeToPlan.chargebeeId)
+                            ? <span>Your account will downgrade to {confirmDowngradeToPlan.name} on the next billing cycle.</span>
+                            : <span>Your account will downgrade to {confirmDowngradeToPlan.name}. The remaining hours in your current plan will be available to use until the next billing cycle.</span>}
+                    </div>
+                </div>
+                <div className="flex justify-end mt-6">
+                    <button className="bg-red-600 border-red-800" onClick={doDowngrade}>Downgrade Plan</button>
+                </div>
+            </Modal>}
         </SettingsPage>
     </div>;
 }
 
+interface PlanCardProps {
+  plan: PlanWithOriginalPrice;
+  isCurrent: boolean;
+  children: React.ReactNode;
+  onUpgrade?: () => void;
+  onDowngrade?: () => void;
+  bottomLabel?: React.ReactNode;
+}
+
+function PlanCard(p: PlanCardProps) {
+    return <SelectableCard className="w-44" title={p.plan.name.toUpperCase()} selected={p.isCurrent} onClick={() => {}}>
+        <div className="mt-5 mb-5 flex flex-col items-center justify-center">
+            <p className="text-3xl text-gray-500 font-bold">{p.plan.hoursPerMonth}</p>
+            <p className="text-base text-gray-500 font-bold">hours</p>
+        </div>
+        <div className="flex-grow flex flex-col space-y-2">{p.children}</div>
+        <div>
+        <p className="text-center text-gray-500 font-semibold mb-2 mt-4">{p.plan.pricePerMonth <= 0.001
+            ? 'FREE'
+            : (p.plan.currency === 'EUR' ? '€' : '$') + p.plan.pricePerMonth + ' per month'
+        }</p>
+        {p.isCurrent
+            ? <button className="w-full" disabled={true}>Current Plan</button>
+            : ((p.onUpgrade && <button className="w-full secondary group-hover:bg-green-600 group-hover:text-gray-100" onClick={p.onUpgrade}>Upgrade</button>)
+                || (p.onDowngrade && <button className="w-full secondary group-hover:bg-green-600 group-hover:text-gray-100" onClick={p.onDowngrade}>Downgrade</button>)
+                || <button className="w-full secondary" disabled={true}>&nbsp;</button>)}
+        </div>
+        <div className="relative w-full">
+            <div className="absolute w-full mt-5 text-center font-semibold">{p.bottomLabel}</div>
+        </div>
+    </SelectableCard>;
+}
+
+function getLocalStorageObject(key: string): PendingPlan | undefined {
+    try {
+        const string = window.localStorage.getItem(key);
+        if (!string) {
+            return;
+        }
+        return JSON.parse(string);
+    } catch (error) {
+        return;
+    }
+}
+
+function removeLocalStorageObject(key: string): void {
+    window.localStorage.removeItem(key);
+}
+
+function setLocalStorageObject(key: string, object: Object): void {
+    window.localStorage.setItem(key, JSON.stringify(object));
```

Comment:
nit: this might throw as well.

Reply:
You mean in case of cyclical references?

Reply:
no, browsers may throw errors. possible reasons: quota, feature disabled, ...